### ISSUE TITLE
Support for multiple Svelte versions?

### ISSUE THREAD

**Issue by saibotsivad** [sveltejs/rollup-plugin-svelte#123]
I've been working with a very sizable business app that's been stuck in Svelte 1 land for... a while.

Because it's reasonably large, transitioning the whole site in one go would be difficult, but it's all in one project so I'm trying to work out ways to transition chunks to Svelte 3.

With npm v6.9.0 you can alias different versions of the same library, so you could do:

```json
"dependencies": {
   "svelte1": "npm:svelte@1",
   "svelte3": "npm:svelte@3"
}
```

However, this plugin puts the Svelte version checking at initialization, e.g. in [index.js](https://github.com/sveltejs/rollup-plugin-svelte/blob/master/index.js) it has:

```js
const { version } = require('svelte/package.json');
// ...
const major_version = +version[0];
// ...
const { compile, preprocess } = major_version >= 3
	? require('svelte/compiler.js')
	: require('svelte');
```

## Proposal?

Imagine if the version of Svelte could be passed in as an option:

```js
import svelte1 from 'svelte1' // through the magic of alias
import svelte3 from 'svelte3'
import sveltePlugin from 'rollup-plugin-svelte'

export default {
  // ...
  plugins: [
    sveltePlugin({
      svelte: svelte1,
      extensions: [ '.html' ]
    }),
    sveltePlugin({
      svelte: svelte3,
      extensions: [ '.svelte' ]
    }),
    // ...
```

It looks like the `compile` and `preprocess` are only used in the returned `transform` function, so this could be a non-breaking change:

```js
function loadSvelte() {
	const { version } = require('svelte/package.json');
	const major_version = +version[0];
	return major_version >= 3
		? require('svelte/compiler.js')
		: require('svelte');
}
module.exports = function svelte(options = {}) {
	const { compile, preprocess } = options.svelte || loadSvelte()
```

I think the general idea is sound, but improvements could probably be made.

**Comment by saibotsivad** [comment]
Note also that I could easily fork this repo and use the fork+changes internally during the transition phase, so no pressure at all if multiple versions doesn't sound like a thing worth maintaining. <3

**Comment by antony** [comment]
I feel like this might be a nice migration path in the future, but I'm less keen on us setting a dependency on npm, or a specific npm version at that.

Perhaps if there was a check for a minimum npm version if you wanted to turn this plugin on, that might be good. Alternatively since you wouldn't be able to do this anyway unless you had the right version of npm, maybe it's ok as is.

I can't say either way right now - but I've actually wanted this feature in the past.

**Comment by lukeed** [comment]
It doesn't require npm – it'd just be one avenue at arriving there. For instance, you can attach a local copy of svelte (for testing/whatever) or reanimate @saibotsivad's example with raw symlinks.

I don't see any issue with this idea – with the sole exception of possible ESM support... if we ever go that direction with this plugin. Although, I guess even then, this plugin would have and import its own `svelte` copy at the top of the file. It'd just do nothing.

```js
// future
import svelte from 'svelte/compiler';

// ...
const compiler = options.svelte || svelte;
```

FWIW, the official `@rollup/typescript` plugin (and a few of the unofficials) do this for custom `typescript` attachments.

**Comment by saibotsivad** [comment]
Yep, looks like [@rollup/typescript does this](https://github.com/rollup/plugins/blob/55a9cb20d7686750aa104d6a45a7d6834f57ba4a/packages/typescript/src/options/plugin.ts#L17):

```ts
import * as defaultTs from 'typescript';
//...
export default function getPluginOptions(options: RollupTypescriptOptions) {
  const { typescript, ...bunchOfOtherStuff } = options;
  return {
    // other stuff
    typescript: typescript || defaultTs
  };
}
``

**Comment by saibotsivad** [comment]
Would it be a good idea for me to make a pull request with these changes, as a more formal proposal? 🤷

**Comment by lukeed** [comment]
Go for it IMO 👍

**Comment by tivac** [comment]
I built a v2-to-v3 translation library, parts of that night be a useful starting point. We used it for a reasonably large long term port of an existing v2 app to v3.

https://github.com/tivac/svelte-translator

Here's some details about the port we did using it.


https://twitter.com/tivac/status/1291426619626647557

It actually started life as a v1 app, we ported to v2 using the automated conversion stuff and that was a relatively painless process.
